### PR TITLE
Add API denylist to Dapr config, complementing API allowlist

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

Dapr has included an API allowlist in the Configuration resource for a number of releases.

However, we've observed that most customers are using the allowlist more as a "denylist", i.e. they list all Dapr APIs except _some_. Feedback we received is that this complicates operations when a new version of Dapr is released.

This PR implements additional support for API denylisting. The denylist is complementary to the allowlist, and the two can be used simultaneously too:

- If only a denylist is defined, all Dapr APIs are allowed except those defined in the denylist
- If only an allowlist is defined, only the Dapr APIs listed in the allowlist are allowed
- If both an allowlist and a denylist are defined, the allowed APIs are those defined in the allowlist _unless_ they are also included in the denylist (i.e. the denylist overrides the allowlist for APIs that are defined in both places).
- If neither is defined, all APIs are allowed.

## Checklist

- [ ] Docs issue: TODO @ItalyPaleAle

### CODE REVIEW

File: pkg/http/server_test.go [hunk 1]
```diff
@@ -58,413 +52,6 @@ func newServer() server {
 	}
 }
 
-func TestAllowedAPISpec(t *testing.T) {
```
Comment (ItalyPaleAle) [inline pkg/http/server_test.go:61]: Reviewers: these tests were moved to the endpoints_test  file (with new tests added)

**Comment by yaron2** [comment]
@ItalyPaleAle please follow Dapr's contribution guidelines going forward and create an issue prior to opening a PR that adds a new feature and changes our specs. Thanks!

**Comment by ItalyPaleAle** [comment]
My apologies, I'll be more mindful next time

**Review by mukundansundar — COMMENTED** [review]
Please add E2E for this feature.

File: charts/dapr/crds/configuration.yaml [hunk 2]
```diff
@@ -85,9 +85,25 @@ spec:
                 description: APISpec describes the configuration for Dapr APIs.
                 properties:
                   allowed:
+                    description: List of allowed APIs. Can be used in conjunction with Denied.
                     items:
-                      description: APIAccessRule describes an access rule for allowing
-                        a Dapr API to be enabled and accessible by an app.
+                      description: APIAccessRule describes an access rule for allowing or disallowing a Dapr API.
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:90]: It comes from the Go struct that's used for both allow and deny so i can't have it different :(

**Comment by ItalyPaleAle** [comment]
> Please add E2E for this feature.

There's currently no E2E test for API allowlisting or denylisting, and although we should create one, perhaps that could be a separate PR?

File: pkg/grpc/endpoints.go [hunk 3]
```diff
@@ -102,33 +102,30 @@ var endpoints = map[string][]string{
...
+					_, ok := allowed[info.FullMethod]
+					if !ok {
+						return nil, invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
+				}
+				if len(denied) > 0 {
+					_, ok := denied[info.FullMethod]
+					if ok {
+						return nil, invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/grpc/endpoints.go:128]: It shouldn't matter. If the same API is both in allowlist and denylist, the method will return an error regardless of the order.

**Comment by daixiang0** [comment]
> > Please add E2E for this feature.
> 
> There's currently no E2E test for API allowlisting or denylisting, and although we should create one, perhaps that could be a separate PR?

Please create an issue to track it.

**Comment by ItalyPaleAle** [comment]
> Please create an issue to track it.

#6240

File: charts/dapr/crds/configuration.yaml [hunk 4]
```diff
@@ -85,9 +85,25 @@ spec:
...
+                          type: string
+                        version:
+                          type: string
+                      required:
+                      - name
+                      - version
+                      type: object
+                    type: array
+                  denied:
+                    description: List of disallowed APIs. Can be used in conjunction with allowed.
```
Comment (daixiang0) [inline charts/dapr/crds/configuration.yaml:104]: typo.

File: charts/dapr/crds/configuration.yaml [hunk 5]
```diff
@@ -85,9 +85,25 @@ spec:
...
+                    items:
+                      description: APIAccessRule describes an access rule for allowing or disallowing a Dapr API.
```
Comment (daixiang0) [inline charts/dapr/crds/configuration.yaml:106]: APIAccessRule describes an access rule to allow or deny one Dapr API.

File: pkg/apis/configuration/v1alpha1/types.go [hunk 6]
```diff
@@ -65,10 +65,15 @@ type ConfigurationSpec struct {
...
+	// List of allowed APIs. Can be used in conjunction with denied.
+	// +optional
 	Allowed []APIAccessRule `json:"allowed,omitempty"`
+	// List of disallowed APIs. Can be used in conjunction with allowed.
+	// +optional
+	Denied []APIAccessRule `json:"denied,omitempty"`
 }
 
-// APIAccessRule describes an access rule for allowing a Dapr API to be enabled and accessible by an app.
+// APIAccessRule describes an access rule for allowing or disallowing a Dapr API.
```
Comment (daixiang0) [inline pkg/apis/configuration/v1alpha1/types.go:76]: allow and deny, let's keep sync

File: pkg/config/configuration.go [hunk 7]
```diff
@@ -139,6 +142,22 @@ type APIAccessRule struct {
 	Protocol string `json:"protocol"`
 }
 
+// APIAccessRules is a list of API access rules (allowlist or denylist).
+type APIAccessRules []APIAccessRule
+
+// ForProtocol returns a list of APIAccessRule objects filtered by protocol
+func (r APIAccessRules) ForProtocol(protocol string) []APIAccessRule {
```
Comment (daixiang0) [inline pkg/config/configuration.go:149]: rename as `GetRulesByProtocol`?

File: pkg/config/configuration.go [hunk 8]
```diff
@@ -139,6 +142,22 @@ type APIAccessRule struct {
...
+	res := make([]APIAccessRule, len(r))
```
Comment (daixiang0) [inline pkg/config/configuration.go:150]: how about `func (r APIAccessRules) ForProtocol(protocol string) (ret []APIAccessRule) {` to reduce `n`?

File: pkg/grpc/endpoints.go [hunk 9]
```diff
@@ -137,11 +134,39 @@ func setAPIEndpointsMiddlewares(rules []config.APIAccessRule) (grpc.UnaryServerI
...
+// Converts a slice of config.APIAccessRule into a map where the key is the gRPC full endpoint
+func apiAccessRuleToMap(rules config.APIAccessRules) map[string]struct{} {
+	res := map[string]struct{}{}
+
+	for _, rule := range rules {
+		if rule.Protocol != "grpc" {
+			continue
+		}
+
+		if list, ok := endpoints[rule.Name+"."+rule.Version]; ok {
```
Comment (daixiang0) [inline pkg/grpc/endpoints.go:164]: Please use const and fmt, also comment or document it for better maintain.

File: pkg/http/server.go [hunk 10]
```diff
@@ -335,8 +335,13 @@ func (s *server) unescapeRequestParametersHandler(next fasthttp.RequestHandler)
 func (s *server) getRouter(endpoints []Endpoint) *routing.Router {
 	router := routing.New()
 	parameterFinder, _ := regexp.Compile("/{.*}")
+
+	// Build the API allowlist and denylist
+	allowedAPIs := s.apiSpec.Allowed.ForProtocol("http")
```
Comment (daixiang0) [inline pkg/http/server.go:340]: So it is better if we make all support protocols as const.

File: pkg/grpc/endpoints.go (hunk 9 above)
Comment (ItalyPaleAle) [reply pkg/grpc/endpoints.go:164]: I will document it, but concatenation is significantly faster than fmt.Sprintf for simple string concatenation:

```go
func BenchmarkFmt(b *testing.B) {
	const name = "foo"
	const version = "bar"
	for i := 0; i < b.N; i++ {
		_ = fmt.Sprintf("%s.%s", name, version)
	}
}
func BenchmarkConcat(b *testing.B) {
	const name = "foo"
	const version = "bar"
	for i := 0; i < b.N; i++ {
		_ = name + "." + version
	}
}
```

On my machine:

```text
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr
BenchmarkFmt
BenchmarkFmt-8      	20474700	        58.38 ns/op	       8 B/op	       1 allocs/op
BenchmarkConcat
BenchmarkConcat-8   	1000000000	         0.3238 ns/op	       0 B/op	       0 allocs/op
```

Not just concatenation is MUCH faster (99.95% faster), but fmt.Sprintf also causes heap allocations so more work on the GC.

File: pkg/apis/configuration/v1alpha1/types.go (hunk 6 above)
Comment (ItalyPaleAle) [reply pkg/apis/configuration/v1alpha1/types.go:76]: Done

File: charts/dapr/crds/configuration.yaml (hunk 4 above)
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:104]: Done

File: pkg/config/configuration.go (hunk 7 above)
Comment (ItalyPaleAle) [reply pkg/config/configuration.go:149]: Done

File: charts/dapr/crds/configuration.yaml (hunk 5 above)
Comment (ItalyPaleAle) [reply charts/dapr/crds/configuration.yaml:106]: Done

File: pkg/http/server.go (hunk 10 above)
Comment (ItalyPaleAle) [reply pkg/http/server.go:340]: Done

File: pkg/grpc/endpoints.go [hunk 11]
```diff
@@ -137,11 +134,40 @@ func setAPIEndpointsMiddlewares(rules []config.APIAccessRule) (grpc.UnaryServerI
...
+					_, ok := allowed[info.FullMethod]
+					if !ok {
+						return invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
+				}
+				if len(denied) > 0 {
+					_, ok := denied[info.FullMethod]
+					if ok {
+						return invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
```
Comment (mukundansundar) [inline pkg/grpc/endpoints.go:147]: @ItalyPaleAle The order here is opposite of the order seeing in `http/endpoint.go`. In the HTTP one, denied is checked first and then allowed is checked. Here the opposite is true.

File: pkg/config/configuration.go [hunk 12]
```diff
@@ -129,14 +129,41 @@ type PipelineSpec struct {
...
+	APIAccessRuleProtocolHTTP APIAccessRuleProtocol = "http"
+	APIAccessRuleProtocolGRPC APIAccessRuleProtocol = "grpc"
+)
+
+// GetRulesByProtocol returns a list of APIAccessRule objects filtered by protocol
+func (r APIAccessRules) GetRulesByProtocol(protocol APIAccessRuleProtocol) []APIAccessRule {
+	res := make([]APIAccessRule, len(r))
+	n := 0
+	for _, v := range r {
+		if v.Protocol == protocol {
```
Comment (mukundansundar) [inline pkg/config/configuration.go:161]: what if the input given is mixed case or all upper case?

**Review by mukundansundar — APPROVED** [review]

**Comment by yaron2** [comment]
Does this have a corresponding docs PR? cc @ItalyPaleAle

**Comment by ItalyPaleAle** [comment]
> Does this have a corresponding docs PR? 

[Now there is.](https://github.com/dapr/docs/pull/3489) Thanks for the reminder.

**Pull request merged**